IMLS/imls_scraper.py

```diff
@@ -0,0 +1,109 @@
+import os
+import sys
+import fnmatch
+from datetime import datetime
+from typing import List, Any
+from zipfile import ZipFile
+from io import BytesIO
+import pandas as pd
+import requests
+from bs4 import BeautifulSoup
+import pandas as pd
+import re
+from pymongo import MongoClient, errors
+from tqdm import tqdm
+import urllib 
+import numpy as np
+
+_i = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+os.environ["DBUSERNAME"] = "democracylab"
+os.environ["PW"] = "DemocracyLab2019"
+if _i not in sys.path:
+    # add parent directory to sys.path so utils module is accessible
+    sys.path.insert(0, _i)
+del _i  # clean up global name space
+from shared_code.utils import (
+    check_similarity, locate_potential_duplicate,
+    insert_services, client
+)
+from shared_code.base_scraper import BaseScraper
+
+     
+class Imls_Scraper(BaseScraper):
+  
+    def __init__(self,
+                 source: str,
+                 data_url: str,
+                 data_page_url: str,
+                 data_format: str,
+                 extract_usecols: List[str],
+                 drop_duplicates_columns: List[str],
+                 rename_columns: dict,
+                 service_summary: Any,
+                 check_collection: str,
+                 dump_collection: str,
+                 dupe_collection: str,
+                 data_source_collection_name: str,
+                 collection_dupe_field: str,
+                 encoding: str = 'utf-8'):
+        BaseScraper.__init__(self, source, data_url, data_page_url, data_format, extract_usecols, drop_duplicates_columns, rename_columns, service_summary, check_collection, dump_collection, dupe_collection, data_source_collection_name, collection_dupe_field, encoding)
+        resp = super().scrape_updated_date()
+        soup = BeautifulSoup(resp, 'html.parser')
+        latestLink = ""
+        latestDate = 0
+        for link in soup.findAll('a', href=True, text='CSV'):
+             match = re.match(r'.*([1-3][0-9]{3})', link['href'])
```

Comment:
can this regular expression be simpler like \d{4} and it can work for any year. Currently it can only work till year 3999.

Reply:
Just wanted to ensure that its for a valid year. \d{4} will even match a string like '0001'. I can increase the range to make it work till year 9999.

Reply:
Sounds good.